internal/blockchain_contracts/src/bitcoin/rfc003/bitcoin_htlc.rs

```diff
@@ -1,137 +1,297 @@
-
-        BitcoinHtlc {
-            script: contract,
-            expiry,
-        }
+        let descriptor = format!(
+            "wsh(c:or_i(and_v(v:sha256({secret_hash}),pk_h({redeem_identity})),and_v(v:older({expiry}),pk_h({refund_identity}))))",
+            secret_hash = secret_hash.to_hex(),
+            redeem_identity = redeem_identity,
+            refund_identity = refund_identity,
+            expiry = expiry,
+        );
+
+        let miniscript =
+            miniscript::Descriptor::from_str(&descriptor).expect("descriptor to be valid");
+
+        BitcoinHtlc { miniscript, expiry }
     }
 
     pub fn compute_address(&self, network: Network) -> Address {
-        Address::p2wsh(&Script::from(self.script.clone()), network)
+        self.miniscript
+            .address(network)
+            .expect("script to be encodable to address")
     }
 
-    pub fn unlock_with_secret(self, secret_key: SecretKey, secret: [u8; 32]) -> UnlockParameters {
-        let public_key = PublicKey::from_secret_key(&*crate::SECP, &secret_key);
-        UnlockParameters {
-            witness: vec![
-                Witness::Signature(secret_key),
-                Witness::PublicKey(public_key),
-                Witness::Data(secret.to_vec()),
-                Witness::Bool(true),
-                Witness::PrevScript,
-            ],
+    pub fn unlock(
+        self,
+        htlc_location: OutPoint,
+        input_value: u64,
+        spend_to: bitcoin::Address,
+        fee_per_wu: u16,
+        strategy: UnlockStrategy,
+    ) -> Result<Transaction, Error> {
+        use UnlockStrategy::*;
+
+        let mut htlc_tx_in = TxIn {
+            previous_output: htlc_location,
+            script_sig: self.miniscript.unsigned_script_sig(),
             sequence: SEQUENCE_ALLOW_NTIMELOCK_NO_RBF,
-            locktime: 0,
-            prev_script: self.into_script(),
+            witness: vec![],
+        };
+        let lock_time = match strategy {
+            Redeem { .. } => 0,
+            Refund { .. } => self.expiry,
+        };
+
+        let mut spending_transaction = Transaction {
+            version: 2,
+            lock_time,
+            input: vec![htlc_tx_in.clone()],
+            output: vec![TxOut {
+                value: 0, // overwritten once we estimated the weight
+                script_pubkey: spend_to.script_pubkey(),
+            }],
+        };
+
+        let base_tx_weight = spending_transaction.get_weight();
+        let output_value = {
+            let expected_weight = base_tx_weight + strategy.expected_witness_stack_weight();
+            // This could potentially overflow but that would mean with have `weight` and
+            // `fee` with absurd
+            let fee = expected_weight * u64::from(fee_per_wu);
+
+            input_value
+                .checked_sub(fee)
+                .ok_or(Error::FeeHigherThanInputValue)?
+        };
+        spending_transaction.output[0].value = output_value;
```

Comment:
Instead of this, let's sign the TX once with 0 output, then calculate the weight and sign it again with the appropriate fee.